jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java

```diff
@@ -0,0 +1,507 @@
+         * <p>
+         * If the vertex $v$ has not been explored, $-1$ will be returned.
+         *
+         * @param v vertex
+         * @return depth of vertex $v$ in the DFS tree/forest
+         */
+        public int getDepth(V v){
+            int index = vertexMap.getOrDefault(v, -1);
+
+            if (index == -1)
+                return -1;
+            else
+                return depth[index];
+        }
+
+        /**
+         * Returns the size of vertex $v$'s subtree in the internal DFS tree/forest.
+         *
+         * <p>
+         * The size of a vertex $v$'s subtree is
+         * defined as the number of vertices in the subtree rooted at $v$ (including $v).
+         *
+         * <p>
+         * If the vertex $v$ has not been explored, $0$ will be returned.
+         *
+         * @param v vertex
+         * @return size of vertex $v$'s subtree in the DFS tree/forest
+         */
+        public int getSizeSubtree(V v){
+            int index = vertexMap.getOrDefault(v, -1);
+
+            if (index == -1)
+                return 0;
+            else
+                return sizeSubtree[index];
+        }
+
+        /**
+         * Returns the component id of vertex $v$ in the internal DFS tree/forest. For two vertices $u$ and $v$,
+         * $component[u] = component[v]$ iff $u$ and $v$ are in the same tree.
+         *
+         * <p>
+         * The component ids are numbers between $0$ and $numberOfTrees - 1$.
+         *
+         * <p>
+         * If the vertex $v$ has not been explored, $-1$ will be returned.
+         *
+         * @param v vertex
+         * @return component id of vertex $v$ in the DFS tree/forest
+         */
+        public int getComponent(V v){
+            int index = vertexMap.getOrDefault(v, -1);
+
+            if (index == -1)
+                return -1;
+            else
+                return component[index];
+        }
+
+        /**
+         * Return the normalized version of the input graph: a map from vertices to unique integers and the reverse
+         * mapping as a list.
+         *
+         * For each vertex $v \in V$, let $vertexMap(v) = x$ such that no two vertices share the same x and all x's are
+         * integers between $0$ and $|V| - 1$. Let $indexList(x) = v$ be the reverse mapping from integers to vertices.
+         *
+         * Note: The two structures returned are immutable.
+         *
+         * @return a pair which consists of the vertexMap and the indexList
+         */
+        public Pair<Map<V, Integer>, List<V>> getNormalizedGraph(){
+            return Pair.of(Collections.unmodifiableMap(vertexMap), Collections.unmodifiableList(indexList));
+        }
+
+        /**
+         * Return a copy of the internal depth array.
+         * For each vertex $v \in V$, $depthArray[normalizeVertex(v)] = getDepth(v)$
+         *
+         * @return internal depth array
+         */
+        public int[] getDepthArray(){
+            return depth.clone();
```

Comment:
We do not expose private fields so we do not really break encapsulation.

Reply:
I think what @d-michail means is that since we're already providing access to internal state for the purpose of efficiency, might was well skip the clone and trust the caller not to modify the original.  I agree, let's skip the clone.

Reply:
But add a comment in the Javadoc, to explicitly explain that.